src/data_structures/interval_tree.rs

```diff
@@ -1,65 +1,118 @@
-//! Interval tree, a data structure for efficiently storing and searching numeric intervals.
+//! Interval tree, a data structure for efficiently storing and searching intervals.
 //!
-//! This data structure uses the `std::ops:Range` type to define the intervals. The range bounds
-//! may be specified by any type from the `num` crate that satisfies the `std::cmp::Ord` trait.
-//! Upon inserting an interval may be associated with a data value. The interval data is stored in
+//! This data structure uses an `Interval` type to define the intervals. It is implemented by
+//! wrapping the `std::ops::Range` in a newtype. An interval must have a positive with and the interval bounds
+//! may be specified by any type satisfies both the `std::cmp::Ord` and `Clone` trait. Because
+//! `Interval` implements `From<Range>` you can also uses normal `Range` arguments in the
+//! `insert` and `find` functions. This implicit conversion will panic if a negative-width range is
+//! supplied.
+//!
+//! Upon inserting an interval may be associated with a data value. The intervals are stored in
 //! an augmented AVL-tree which allows for efficient inserting and querying.
 //!
 //! # Example
 //! ```
-//! use bio::data_structures::interval_tree::IntervalTree;
+//! use bio::data_structures::interval_tree::{IntervalTree, Interval};
 //!
 //! let mut tree = IntervalTree::new();
-//! tree.insert((11..20), "Range_1").unwrap();
-//! tree.insert((25..30), "Range_2").unwrap();
-//! for r in tree.find(&(15..25)).unwrap() {
-//!     assert_eq!(r.interval(), &(11..20));
+//! tree.insert(11..20, "Range_1");
+//! tree.insert(25..30, "Range_2");
+//! for r in tree.find(15..25) {
+//!     assert_eq!(r.interval().start, 11);
+//!     assert_eq!(r.interval().end, 20);
+//!     assert_eq!(r.interval(), &(Interval::from(11..20)));
 //!     assert_eq!(r.data(), &"Range_1");
 //! }
-extern crate num;
-
-use self::num::traits::Num;
+//! ```
+//!
 
 use std::cmp;
 use std::mem;
-use std::fmt::Debug;
-use std::ops::Range;
+use std::ops::{Range, Deref};
+use std::iter::FromIterator;
 
-/// An interval tree for storing Ranges with data
+/// An interval tree for storing intervals with data
 #[derive(Debug, Clone)]
-pub struct IntervalTree<N, D> {
+pub struct IntervalTree<N: Ord + Clone, D> {
     root: Option<Node<N, D>>,
 }
 
-/// An `Entry` is used by the IntervalTreeIterator to return references to the data in the tree
+/// An `Interval` wraps the `std::ops::Range` from the stdlib and is defined by a start and end field
+/// where end should be >= start.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Interval<N: Ord + Clone>(Range<N>);
+
+impl<N: Ord + Clone> Interval<N> {
+    /// Construct a new `Interval` from the given Range.
+    /// Will return `Err` if end < start.
+    pub fn new(r: Range<N>) -> Result<Interval<N>, IntervalTreeError> {
+        if r.end >= r.start {
+            Ok(Interval(r))
+        } else {
+            Err(IntervalTreeError::InvalidRange)
```

Comment:
Can you please rename the IntervalTreeError into an IntervalError? This makes Intervals usable in other places of the Rust-Bio as well. Also, would you be fine to move them into utils.rs?